### ISSUE TITLE
Error: Anomaly "Uncaught exception Not_found." Please report at http://coq.inria.fr/bugs/.

### ISSUE THREAD

**Issue by roglo** [coq/coq#19587]
### Description of the problem

While compiling my application with Coq 8.20.0. Seems to also fail with 8.18.0. Hard to isolate the problem. Please clone "git@github.com:roglo/puiseuxth.git", branch "coq-8.20.0", go to directory "coq", type "make". It fails at file F1Eq.v.

### Version of Coq where this bug occurs

8.20.0

### Last version of Coq where the bug did not occur

Also failing at 8.18.0, perhaps before.

**Comment by SkySkimmer** [comment]
@coqbot: minimize
```bash
#!/usr/bin/env bash
git clone "https://github.com/roglo/puiseuxth.git" -b coq-8.20.0
cd puiseuxth/coq
make
```

**Comment by JasonGross** [comment]
There was an issue with empty arguments, let's try again after https://github.com/coq-community/run-coq-bug-minimizer/commit/285d57d73fd4f9571b59eabe91902587f6074f02
@coqbot: minimize
```bash
#!/usr/bin/env bash
git clone "https://github.com/roglo/puiseuxth.git" -b coq-8.20.0
cd puiseuxth/coq
make
```

**Comment by JasonGross** [comment]
I guess the runner fails when coqc is loading files from pwd implicitly.  That might be fixed by https://github.com/coq-community/run-coq-bug-minimizer/pull/35, but in the meantime, let's try
@coqbot: minimize
```bash
#!/usr/bin/env bash
git clone "https://github.com/roglo/puiseuxth.git" -b coq-8.20.0
cd puiseuxth/coq
make || true
coqc -q -Q . "" F1Eq.v
```

**Comment by JasonGross** [comment]
Apparently it also does not work when the final argument is the empty string
@coqbot: minimize
```bash
#!/usr/bin/env bash
git clone "https://github.com/roglo/puiseuxth.git" -b coq-8.20.0
cd puiseuxth/coq
make || true
coqc -Q . "" -q F1Eq.v
```

**Comment by roglo** [comment]
Not sure it helps, but I reduced the bug from your version to 240 lines. I am trying to see if I can reduce it more. See https://github.com/roglo/puiseuxth/blob/coq-8.20.0/coq/Bug.v

**Comment by ppedrot** [comment]
Slightly further reduced:
```coq
(* coqc version 8.20.0 compiled with OCaml 4.13.1
   coqtop version 8.20.0
   Expected coqc runtime on this file: 0.352 sec *)

Require Coq.QArith.QArith.

Set Implicit Arguments.

Class ring α := { rng_zero : α; }.

Class field α (rng_ring : ring α) := { fld_inv : α -> α }.

Import Coq.QArith.QArith.

Inductive lap_eq {α} {r : ring α} : list α -> list α -> Prop :=
  | lap_eq_cons : forall x₁ x₂ l₁ l₂, lap_eq (x₁ :: l₁) (x₂ :: l₂).

Theorem lap_eq_refl {α} {r : ring α} : reflexive _ lap_eq.
Admitted.

Theorem lap_eq_sym {α} {r : ring α} : symmetric _ lap_eq.
Admitted.

Theorem lap_eq_trans {α} {r : ring α} : transitive _ lap_eq.
Admitted.

Add Parametric Relation α (r : ring α) : (list α) lap_eq
 reflexivity proved by lap_eq_refl
 symmetry proved by lap_eq_sym
 transitivity proved by lap_eq_trans
 as lap_eq_rel.

Definition lap_mul {α} {R : ring α} (la lb : list α) : list α := nil.

Global Instance lap_mul_morph α (R : ring α) :
  Proper (lap_eq ==> lap_eq ==> lap_eq) (@lap_mul _ R).
Admitted.

Theorem lap_eq_0 : forall (α : Type) (r : ring α), lap_eq (cons rng_zero nil) nil.
Admitted.

Definition series_0 {α} {r : ring α} : nat -> α := fun i => rng_zero.

Record puiseux_series α := mkps
  { ps_terms : nat -> α }.

Definition ps_zero {α} {r : ring α} :=
  {| ps_terms := series_0 |}.

Definition ps_ring α (R : ring α) (K : field R) : ring (puiseux_series α) := {| rng_zero := ps_zero; |}.

Canonical Structure ps_ring.

Theorem glop : forall
  (α : Type) (R : ring α) (K : field R),
  @lap_eq (puiseux_series α) (@ps_ring α R K)
     (@lap_mul (puiseux_series α) (@ps_ring α R K)
        (@cons (puiseux_series α) (@ps_zero α R) (@nil (puiseux_series α)))
        nil
     )
     nil.
Proof.
intros.
Check 1%nat.
rewrite lap_eq_0.
Check 1%nat.
```

**Comment by roglo** [comment]
You can also remove lap_mul and lap_mul_morph

Theorem glop :
  @lap_eq (puiseux_series α) (@ps_ring α R K)
     (@cons (puiseux_series α) (@ps_zero α R) (@nil (puiseux_series α)))
     [].

**Comment by roglo** [comment]
No error at coq 8.16.1. Fails at coq 8.18.0.

**Comment by ppedrot** [comment]
Further reduced:
```coq
Require Import Coq.Setoids.Setoid Coq.Classes.Morphisms.

Set Implicit Arguments.

Class ring α := { rng_zero : α; }.
Class field α (rng_ring : ring α) := { }.

Parameter lap_eq : forall {α} {r : ring α}, list α -> list α -> Prop.

Declare Instance  lap_eq_equiv : forall {α} {r : ring α}, Equivalence lap_eq.

Axiom lap_eq_0 : forall (α : Type) (r : ring α), lap_eq (cons rng_zero nil) nil.

Definition puiseux_series (α : Type) := nat -> α.

Definition ps_zero {α} {r : ring α} : puiseux_series α := fun i => rng_zero.

Definition ps_ring α (R : ring α) (K : field R) : ring (puiseux_series α) :=
  {| rng_zero := ps_zero; |}.

Canonical Structure ps_ring.

Theorem glop : forall
  (α : Type) (R : ring α) (K : field R),
  @lap_eq (puiseux_series α) (@ps_ring α R K)
     (@cons (puiseux_series α) (@ps_zero α R) (@nil (puiseux_series α)))
     nil.
Proof.
intros.
Check 1%nat.
rewrite lap_eq_0.
Check 1%nat.
```

**Comment by ejgallego** [comment]
The script is pretty good for bisecting @ppedrot , I can launch an automatic bisect if you want.

[Setting as bisect command `make dunestrap && dune build theories/Foo/foo.v` works pretty well for example]

**Comment by ejgallego** [comment]
My bisect points at 3a6ba81077c81256c10aa04b542535c920b032c4

**Comment by roglo** [comment]
No need to "Set Implicit Arguments", nor type parameters for "field"

```
Require Import Coq.Setoids.Setoid Coq.Classes.Morphisms.

Class ring α := { rng_zero : α; }.
Class field := { }.

Parameter lap_eq : forall {α} {r : ring α}, list α -> list α -> Prop.

Declare Instance lap_eq_equiv : forall {α} {r : ring α}, Equivalence lap_eq.

Axiom lap_eq_0 : forall (α : Type) (r : ring α), lap_eq (cons rng_zero nil) nil.

Definition puiseux_series (α : Type) := nat -> α.

Definition ps_zero {α} {r : ring α} : puiseux_series α := fun i => rng_zero.

Definition ps_ring α (R : ring α) (K : field) : ring (puiseux_series α) :=
  {| rng_zero := ps_zero; |}.

Canonical Structure ps_ring.

Theorem glop : forall
  (α : Type) (R : ring α) (K : field),
  @lap_eq (puiseux_series α) (@ps_ring α R K)
     (@cons (puiseux_series α) (@ps_zero α R) (@nil (puiseux_series α)))
     nil.
Proof.
intros.
Check 1%nat.
rewrite lap_eq_0.
Check 1%nat.
```

**Comment by roglo** [comment]
Good at Coq 8.17.1
Bad at Coq 8.18.0

**Comment by ppedrot** [comment]
Seems like we're not properly removing defined evars from the set of TC evars in the evarmap.

**Comment by ppedrot** [comment]
Fixed by #19594
